spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,426 @@
+package org.apache.spark.sql.cassandra
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.catalyst.expressions.{Literal, AttributeReference, Expression, Attribute}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.{ReadConf, ValidRDDType}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+/**
+ * Table scan type. It allows to create a Relation for given table, keyspace, cluster and user
+ * defined schema. It also allow to pass [[CassandraConnector]], [[ReadConf]] and [[WriteConf]]
+ * explicitly
+ */
+sealed trait ScanType {
```

Comment:
First, not sure why you need sealed / case objects here. Are you ever pattern matching on them? 

Second, this ScanType hierarchy is solely used as factories. Every ScanType produces an instance of `BaseRelation` with `InsertableRelation`, and each type of scan has a corresponding relation implementation. This way you have two parllel class hierarchies. 

A standard way of doing such thing in Scala is to use companion objects as factories. This implies the same name for a factory object and its implementation class. 

Therefore you'd have:
```scala

// in CassandraRelation.scala:
class CassandraRelation extends BaseRelation with InsertableRelation with TableScan
object CassandraRelation extends ScanType

// in PrunedCassandraRelation.scala:
class PrunedCassandraRelation extends CassandraRelation with InsertableRelation with PrunedScan
object PrunedCassandraRelation extends ScanType

// in PrunedFilteredCassandraRelation
class PrunedFilteredCassandraRelation extends ... with PrunedFilteredScan
object PrunedFilteredCassandraRelation extends ScanType

... // similar for Catalyst
```
Let's go further, do we really need 4 different relation types? There is lot of code duplication in the hierarchy. Can't we just implement a single `CassandraRelation` that can do all the pushing down / filtering and implement all the required Spark traits (TableScan, PrunedScan, PrunedFilteredScan)?